### PR TITLE
Hook up registration/login APIs and implement access token generation

### PR DESCRIPTION
**Pull request by Kegsay**
This finishes feature work on registration/login for now.

Registration:
```
$ curl -XPOST -d '{"username":"alice","password":"12345678"}' "http://localhost:7778/api/_matrix/client/r0/register"
{
	"flows": [{
		"stages": ["m.login.dummy"]
	}],
	"completed": [],
	"params": {},
	"session": "totallyuniquesessionid"
}

$ curl -XPOST -d '{"username":"alice","password":"12345678","auth":{"type":"m.login.dummy"}}' "http://localhost:7778/api/_matrix/client/r0/register"
{
	"user_id": "@alice:localhost",
	"access_token": "mV8NAhs90FwKISo5_dlSVqxnqn0AYkS11YnW4faVLxc",
	"home_server": "localhost",
	"device_id": "unknown-device"
}

$ curl -XPOST -d '{"username":"bob","password":"1","auth":{"type":"m.login.dummy"}}' "http://localhost:7778/api/_matrix/client/r0/register"
{"errcode":"M_WEAK_PASSWORD","error":"password too weak: min 8 chars"}
```

Login:
```
$ curl -XGET  "http://localhost:7778/api/_matrix/client/r0/login"
{
	"flows": [{
		"type": "m.login.password",
		"stages": ["m.login.password"]
	}]
}

$ curl -XPOST -d '{"user":"alice","password":"invalid","type":"m.login.password"}' "http://localhost:7778/api/_matrix/client/r0/login"
{"errcode":"M_BAD_JSON","error":"username or password was incorrect, or the account does not exist"}

$ curl -XPOST -d '{"user":"alice","password":"12345678","type":"m.login.password"}' "http://localhost:7778/api/_matrix/client/r0/login"
{
	"user_id": "@alice:localhost",
	"access_token": "d_4_PLGDSRdVXv4-r0b1xhU8VGhNvV1LcsYkx2nOLLI",
	"home_server": "localhost"
}

$ curl -XPOST -d '{}' "http://localhost:7778/api/_matrix/client/r0/createRoom?access_token=@alice:localhost"
{"errcode":"M_FORBIDDEN","error":"Invalid access token"}

$ curl -XPOST -d '{}' "http://localhost:7778/api/_matrix/client/r0/createRoom?access_token=d_4_PLGDSRdVXv4-r0b1xhU8VGhNvV1LcsYkx2nOLLI"
{ .... }
```

Repeated logins clobber access tokens for now (so there is only ever 1 valid access_token at any one time, meaning multi-device doesn't work). This is keyed off the device ID, which we don't yet parse out the request.

### CODE REVIEW

File: src/github.com/matrix-org/dendrite/clientapi/auth/auth.go [hunk 1]
```diff
@@ -27,6 +29,13 @@ import (
 	"github.com/matrix-org/util"
 )
 
+// UnknownDeviceID is the default device id if one is not specified.
+var UnknownDeviceID = "unknown-device"
```
Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/auth/auth.go]: Is this specified somewhere?
Is it okay for multiple clients to share a device ID?

Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/auth/auth.go]: It looks like synapse generates a random device ID if one isn't specified?
https://github.com/matrix-org/synapse/blob/master/synapse/handlers/device.py#L82

Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/auth/auth.go]: > Repeated logins clobber access tokens for now (so there is only ever 1 valid access_token at any one time, meaning multi-device doesn't work). This is keyed off the device ID, which we don't yet parse out the request.

I'm aware this deviates from Synapse for now. I would rather we didn't amass a huge list of *valid* access tokens for an account, so I'm limiting it to 1 unknown device for now. Clients should be giving us device IDs.

Comment (NegativeMjark) [inline src/github.com/matrix-org/dendrite/clientapi/auth/auth.go]: Seems reasonable. Could you add a comment explaining this?

**Comment by NegativeMjark** [comment]
Other than commenting on device ID. LGTM

File: src/github.com/matrix-org/dendrite/clientapi/auth/auth.go (hunk 1 above)
Comment (Kegsay) [inline src/github.com/matrix-org/dendrite/clientapi/auth/auth.go]: Done. https://github.com/matrix-org/dendrite/pull/122/commits/1f8da07c416b4b97d1ca259ea3eb1605aca1f96c

**Pull request merged**